tests/contracts/wallet_2.scilla

```diff
@@ -0,0 +1,425 @@
+        e = mk_transaction_signed_event;
+        event e
+      end
+    end
+  end
+end
+
+(* Execute signed-off transaction *)
+transition ExecuteTransaction (transactionId : Uint32)
+  transaction_opt <- transactions[transactionId];
+  match transaction_opt with
+  | None =>
+    (* Transaction was not found. *)
+    err = UnknownTransactionId;
+    e = mk_error_event err;
+    event e
+  | Some (Trans recipient amount tag) =>
+    (* Only the recipient or an owner can execute the transaction *)
+    recipient_is_sender = builtin eq recipient _sender;
+    sender_is_owner <- exists owners[_sender];
+    sender_may_execute = orb recipient_is_sender sender_is_owner;
+    match sender_may_execute with
+    | False =>
+      err = SenderMayNotExecute;
+      e = mk_error_event err;
+      event e
+    | True =>
+      (* Check for sufficient funds  *)
+      bal <- _balance;
+      not_enough_money = builtin lt bal amount;
+      match not_enough_money with
+      | True =>
+        err = InsufficientFunds;
+        e = mk_error_event err;
+        event e
+      | False =>
+        sigs_opt <- signatures[transactionId];
+        match sigs_opt with
+        | None =>
+          (* Signatures not found, even though the transaction exists.*)
+          err = NoSignatureListFound;
+          e = mk_error_event err;
+          event e
+        | Some sigs =>
+          (* Check for sufficient number of signatures *)
+          no_of_sigs = builtin size sigs;
+          not_enough_signatures = builtin lt no_of_sigs required_signatures;
+          match not_enough_signatures with
+          | True =>
+            err = NotEnoughSignatures;
+            e = mk_error_event err;
+            event e
+          | False =>
+            (* Transaction approved, and enough money available. *)
+            (* Remove transaction and signatures, and execute. *)
+            delete transactions[transactionId];
+            delete signatures[transactionId];
+            msgs = transaction_msg_as_list recipient amount tag;
+            send msgs
+          end
+        end
+      end
+    end
+  end
+end
+
+(* Revoke signature of existing transaction, if it has not yet been executed. *)
+transition RevokeSignature (transactionId : Uint32)
+  (* Sender must have signed already, and transaction must not have been executed. *)
+  sender_has_signed <- exists signatures[transactionId][_sender];
+  match sender_has_signed with
+  | False =>
+    err = NotAlreadySigned;
+    e = mk_error_event err;
+    event e
+  | True =>
+    delete signatures[transactionId][_sender]
+  end
+end
+
+(* Add funds to wallet *)
+transition AddFunds ()
+  (* Check validity of contract. If the contract is invalid, funds will not be accepted. *)
+  validity <- contract_valid;
+  match validity with
+  | Unchecked =>
+    (* Check validity and build owners map *)
+    owners_map_opt = check_validity_and_build_owners_map owners_list required_signatures;
```

Comment:
Can we not do `check_validity_and_build_owners_map` when initializing `owners`? so that if it is found to be invalid, an error / exception can be thrown, thus preventing such a contract from being deployed at all.

1. This prevents dead contracts being deployed (blockchain storage space not wasted).
2. The check for validity on every run of `AddFunds()` can be simplified (or eliminated), thus saving gas.

Reply:
Hmm... `throw` is a statement, so we can't use that in a field initializer. 

I can divide by 0 instead...? It's not nice, though.

Alternatively, I can let `owners_map` be an optional map, and let `None` indicate that the contract is invalid?

Reply:
We have a situation where we want a throw expression! I think we didn't implement it because there wasn't a use case (@anton).

How about an Issue to introduce throw expressions, link it to this and use either of the two alternate solutions you propose? (I'm OK with either).